Source/MQTTnet.AspnetCore/ReaderExtensions.cs

```diff
@@ -81,23 +83,24 @@ static bool TryReadBodyLength(ref ReadOnlySequence<byte> input, out int headerLe
             var index = 1;
             headerLength = 0;
             bodyLength = 0;
-            
-            var temp = input.Slice(0, Math.Min(5, input.Length)).GetMemory().Span;
+
+            var valueSequence = input.Slice(0, Math.Min(5, input.Length));
+            var valueSpan = valueSequence.IsSingleSegment ? valueSequence.First.Span : valueSequence.ToArray();
 
             do
             {
-                if (index == temp.Length)
+                if (index == valueSpan.Length)
                 {
                     return false;
                 }
 
-                encodedByte = temp[index];
+                encodedByte = valueSpan[index];
                 index++;
 
                 value += (byte)(encodedByte & 127) * multiplier;
                 if (multiplier > 128 * 128 * 128)
                 {
-                    throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", temp.Slice(1, index).ToArray())}).");
+                    throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", valueSpan.Slice(1, index).ToArray())}).");
```

Comment:
This check should be moved down after L106 where `multiplier` is updated.
Then as this materializes a costly string it should use a throw-helper. Like
```c#
// L105
multiplier *= 128;

if (multiplier > 128 * 128 * 128)
{
    Throw(valueSpan, index);

    static void Throw(ReadOnlySpan<byte> valueSpan, int index)
        => throw new MqttProtocolViolationException($"Remaining length is invalid (Data={string.Join(",", valueSpan.Slice(1, index).ToArray())}).");
}
```
This keeps the loop-body short (from CPU-view), thus faster.

Reply:
Great!!